apollo-federation/src/sources/connect/validation/selection.rs

```diff
@@ -111,89 +122,175 @@ struct SelectionValidator<'schema> {
+        };
+        let Some(ty) = self
+            .schema
+            .get_object(field.definition.ty.inner_named_type())
+        else {
+            return Ok(None);
+        };
+        Ok(Some(Group {
+            selection,
+            ty,
+            definition: field.definition,
+        }))
+    }
+
+    /// Get all the fields for an object type / selection.
+    /// Returns an error if a selection points at a field which does not exist on the schema.
+    fn enter_group(&mut self, group: &Group<'schema>) -> Result<Vec<Field<'schema>>, Self::Error> {
+        self.path.push(PathPart::Field {
+            definition: group.definition,
+            ty: group.ty,
+        });
+        group.selection.selections_iter().map(|selection| {
+            let field_name = selection.name();
+            let definition = group.ty.fields.get(field_name).ok_or_else(|| {
+                Message {
+                    code: Code::SelectedFieldNotFound,
+                    message: format!(
+                        "{coordinate} contains field `{field_name}`, which does not exist on `{parent_type}`.",
+                        coordinate = &self.selection_coordinate,
+                        parent_type = group.ty.name,
+                    ),
+                    locations: self.selection_location.iter().cloned().collect(),
+                }
+            })?;
+            Ok(Field {
+                selection,
+                definition,
+            })
+        }).collect()
+    }
 
     fn exit_group(&mut self) -> Result<(), Self::Error> {
         self.path.pop();
         Ok(())
     }
+}
 
-    fn finish(self) -> Result<(), Self::Error> {
-        Ok(())
+impl<'schema> FieldVisitor<Field<'schema>> for SelectionValidator<'schema> {
+    type Error = Message;
+
+    fn visit(&mut self, field: Field<'schema>) -> Result<(), Self::Error> {
+        let field_name = field.definition.name.as_str();
+        let type_name = field.definition.ty.inner_named_type();
+        let field_type = self.schema.types.get(type_name).ok_or_else(|| Message {
+            code: Code::GraphQLError,
+            message: format!(
+                "{coordinate} contains field `{field_name}`, which has undefined type `{type_name}.",
+                coordinate = &self.selection_coordinate,
+            ),
+            locations: self.selection_location.iter().cloned().collect(),
+        })?;
+        let is_group = field.selection.next_subselection().is_some();
+
+        match (field_type, is_group) {
+            (ExtendedType::Object(object), true) => {
+                self.check_for_circular_reference(field, object)
+            },
+            (_, true) => {
+                Err(Message {
+                    code: Code::GroupSelectionIsNotObject,
+                    message: format!(
+                        "{coordinate} selects a group `{field_name}{{}}`, but `{parent_type}.{field_name}` is of type `{type_name}` which is not an object.",
```

Comment:
Is the extra `{{}}` needed here?

Reply:
I want to have the braces to better indicate to users what we mean by "group", since that's not a term they'll necessarily be familiar with. So the message would be like "selections a group `user {}`", evoking the selection syntax.